rhino/src/main/java/org/mozilla/javascript/SlotMapOwner.java

```diff
@@ -0,0 +1,60 @@
+package org.mozilla.javascript;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+
+public abstract class SlotMapOwner {
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * This holds all the slots. It may or may not be thread-safe, and may expand itself to a
+     * different data structure depending on the size of the object.
+     */
+    private volatile SlotMap slotMap;
```

Comment:
I'm a bit concerned about making this volatile by default given how frequently it is accessed. Reading the docs for VarHandle, though, it looks like we can use it to choose volatile access semantics when we need them (in the thread-safe case) and non-volatile semantics otherwise. Do you know if that's something that we can do? I would be more comfortable with this if I could try to reduce as much overhead as possible in the single-threaded case.

Reply:
I was expecting a question along those lines whether I made it volatile or not. I made it volatile as that was an easier way to think about things, but I believe it can be made non-volatile since all the multithreaded writes should be protected by fences. I'd like to think about this carefully before changing it though, and probably go through the §17.4 of the JLS and the VarHandles docs to make sure we aren't leaving any holes in the reasoning.

Reply:
Yes, I'd love to see if we can make this work with the same semantics without making that field volatile -- let me know what you learn about whether we can do that!

Reply:
Okay. I'm happy that making the field non-volatile is okay. All access in multithreaded environments is either to immutable maps which are replaced with a CAS operation, or behind a lock.